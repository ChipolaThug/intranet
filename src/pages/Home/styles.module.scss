@import "../../theme/mixins";

.homePage {
    @include page-wrapper-center;
}

.homeContent {
    @include page-content-wrapper(600px);
}
