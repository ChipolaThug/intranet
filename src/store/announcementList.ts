import { PayloadAction, createSlice, createEntityAdapter } from "@reduxjs/toolkit"

import { StateRequest, toastError, elementListFetch } from "./utils"
import { Announcement } from "../services/announcement"
import { AppThunk, AppState } from "."
import { announcementListGet } from "../services/announcementAccessors"

const announcementAdapter = createEntityAdapter<Announcement>()

export const initialState = announcementAdapter.getInitialState({
    readyStatus: "idle",
} as StateRequest)

const announcementList = createSlice({
    name: "announcementList",
    initialState,
    reducers: {
        getRequesting: (state) => {
            state.readyStatus = "request"
        },
        getSuccess: (state, { payload }: PayloadAction<Announcement[]>) => {
            state.readyStatus = "success"
            announcementAdapter.setAll(state, payload)
        },
        getFailure: (state, { payload }: PayloadAction<string>) => {
            state.readyStatus = "failure"
            state.error = payload
        },
    },
})

export default announcementList.reducer
export const { getRequesting, getSuccess, getFailure } = announcementList.actions

export const fetchAnnouncementList = elementListFetch(
    announcementListGet,
    getRequesting,
    getSuccess,
    getFailure,
    (error: Error) => toastError(`Erreur lors du chargement des announcements: ${error.message}`)
)

const shouldFetchAnnouncementList = (state: AppState) =>
    state.announcementList.readyStatus !== "success"

export const fetchAnnouncementListIfNeed = (): AppThunk => (dispatch, getState) => {
    const { jwt } = getState().auth
    if (shouldFetchAnnouncementList(getState())) return dispatch(fetchAnnouncementList(jwt))

    return null
}
