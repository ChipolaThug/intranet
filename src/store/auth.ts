import { createSelector, createSlice, PayloadAction } from "@reduxjs/toolkit"
import { AppState } from "."

// Define a type for the slice state
interface AuthState {
    jwt: string
    id: number
    roles: string[]
}

// Define the initial state using that type
const initialState: AuthState = {
    jwt: "",
    id: 0,
    roles: [],
}

export const auth = createSlice({
    name: "auth",
    initialState,
    reducers: {
        setCurrentUser: (state, action: PayloadAction<AuthState>) => {
            state.jwt = action.payload.jwt
            state.id = action.payload.id
            state.roles = action.payload.roles
        },
        logoutUser: (state) => {
            // Unused, just reload page :/
            state.jwt = ""
            state.id = 0
            state.roles = []
        },
    },
})

export const { setCurrentUser, logoutUser } = auth.actions

const selectAuthData = (state: AppState): AuthState => state.auth

const selectRouter = (state: AppState): AppState["router"] => state.router

export const selectUserJwtToken = createSelector(selectAuthData, (authData) => authData.jwt)

export const routerSelector = createSelector(selectRouter, (authData) => authData)

export const selectUserRoles = createSelector(selectAuthData, (authData) => authData.roles)

export const selectUserId = createSelector(selectAuthData, (authData) => authData.id)

export const isUserConnected = createSelector(selectUserJwtToken, (token) => !!token)

export default auth.reducer
