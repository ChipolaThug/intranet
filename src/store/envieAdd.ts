import { PayloadAction, createSlice, createEntityAdapter } from "@reduxjs/toolkit"
import { toast } from "react-toastify"

import { StateRequest } from "./utils"
import { Envie, EnvieWithoutId, addEnvie } from "../services/envies"
import { AppThunk } from "."

const envieAdapter = createEntityAdapter<Envie>({
    selectId: (envie) => envie.envieId,
})

const envieAdd = createSlice({
    name: "addEnvie",
    initialState: envieAdapter.getInitialState({
        readyStatus: "idle",
    } as StateRequest),
    reducers: {
        getRequesting: (state) => {
            state.readyStatus = "request"
        },
        getSuccess: (state, { payload }: PayloadAction<Envie>) => {
            state.readyStatus = "success"
            envieAdapter.addOne(state, payload)
        },
        getFailure: (state, { payload }: PayloadAction<string>) => {
            state.readyStatus = "failure"
            state.error = payload
        },
    },
})

export default envieAdd.reducer
export const { getRequesting, getSuccess, getFailure } = envieAdd.actions

export const postEnvie =
    (envieWithoutId: EnvieWithoutId): AppThunk =>
    async (dispatch) => {
        dispatch(getRequesting())

        const { error, data } = await addEnvie(envieWithoutId)

        if (error) {
            dispatch(getFailure(error.message))
            toast.error(`Erreur lors de l'ajout d'une envie: ${error.message}`, {
                position: "top-center",
                autoClose: 6000,
                hideProgressBar: true,
                closeOnClick: true,
                pauseOnHover: true,
                draggable: true,
                progress: undefined,
            })
        } else {
            dispatch(getSuccess(data as Envie))
            toast.success("Envie ajout√©e !", {
                position: "top-center",
                autoClose: 3000,
                hideProgressBar: true,
                closeOnClick: true,
                pauseOnHover: true,
                draggable: true,
                progress: undefined,
            })
        }
    }
