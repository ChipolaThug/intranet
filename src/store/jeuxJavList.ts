import { PayloadAction, createSlice } from "@reduxjs/toolkit"

import { JeuxJav, getJeuxJavList } from "../services/jeuxJav"
import { AppThunk, AppState } from "."

interface JeuxJavList {
    readyStatus: string
    items: JeuxJav[]
    error: string | null
}

export const initialState: JeuxJavList = {
    readyStatus: "invalid",
    items: [],
    error: null,
}

const jeuxJavList = createSlice({
    name: "jeuxJavList",
    initialState,
    reducers: {
        getRequesting: (state: JeuxJavList) => {
            state.readyStatus = "request"
        },
        getSuccess: (state, { payload }: PayloadAction<JeuxJav[]>) => {
            state.readyStatus = "success"
            state.items = payload
        },
        getFailure: (state, { payload }: PayloadAction<string>) => {
            state.readyStatus = "failure"
            state.error = payload
        },
    },
})

export default jeuxJavList.reducer
export const { getRequesting, getSuccess, getFailure } = jeuxJavList.actions

export const fetchJeuxJavList = (): AppThunk => async (dispatch) => {
    dispatch(getRequesting())

    const { error, data } = await getJeuxJavList()

    if (error) {
        dispatch(getFailure(error.message))
    } else {
        dispatch(getSuccess(data as JeuxJav[]))
    }
}

const shouldFetchJeuxJavList = (state: AppState) => state.jeuxJavList.readyStatus !== "success"

export const fetchJeuxJavListIfNeed = (): AppThunk => (dispatch, getState) => {
    if (shouldFetchJeuxJavList(getState())) return dispatch(fetchJeuxJavList())

    return null
}
