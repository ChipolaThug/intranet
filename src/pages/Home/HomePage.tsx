import { FC, memo } from "react"
import { RouteComponentProps } from "react-router-dom"
import { Helmet } from "react-helmet"

import { AppThunk } from "../../store"
import styles from "./styles.module.scss"

export type Props = RouteComponentProps

const fetchUserData = () => () => Promise.resolve()

const HomePage: FC<Props> = (): JSX.Element => (
    <div className={styles.homePage}>
        <div className={styles.homeContent}>
            <Helmet title="Home" />
            <div>Tableau de bord</div>
        </div>
    </div>
)

// Fetch server-side data here
export const loadData = (): AppThunk[] => [
    fetchUserData(),
    // More pre-fetched actions...
]

export default memo(HomePage)
