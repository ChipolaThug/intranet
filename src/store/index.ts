import { createMemoryHistory, createBrowserHistory } from "history"
import { Action, configureStore, EntityState } from "@reduxjs/toolkit"
import { ThunkAction } from "redux-thunk"
import { routerMiddleware } from "connected-react-router"
import Cookies from "js-cookie"

import createRootReducer from "./rootReducer"
import { StateRequest } from "./utils"
import { setCurrentUser, logoutUser } from "./auth"

interface Arg {
    initialState?: typeof window.__INITIAL_STATE__
    url?: string
    jwt?: string
    id?: number
}

// Use inferred return type for making correctly Redux types
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
const createStore = ({ initialState, url, jwt, id }: Arg = {}) => {
    const history = __SERVER__
        ? createMemoryHistory({ initialEntries: [url || "/"] })
        : createBrowserHistory()
    const store = configureStore({
        preloadedState: initialState,
        reducer: createRootReducer(history),
        middleware: (getDefaultMiddleware) => [
            // Included default middlewares: https://redux-toolkit.js.org/api/getDefaultMiddleware#included-default-middleware
            ...getDefaultMiddleware(),
            routerMiddleware(history),
        ],
        devTools: __DEV__,
    })

    if (jwt && id) {
        store.dispatch(setCurrentUser({ jwt, id }))
    } else {
        store.dispatch(logoutUser())
    }

    return { store, history }
}

const storage: any = localStorage
const id = +(Cookies.get("id") || storage?.getItem("id"))
const jwt = Cookies.get("jwt") || storage?.getItem("jwt")

const { store } = createStore({ id, jwt })

export type AppState = ReturnType<typeof store.getState>

export type AppDispatch = typeof store.dispatch

export type AppThunk = ThunkAction<void, AppState, unknown, Action<string>>

export type EntitiesRequest<T> = EntityState<T> & StateRequest

export type ValueRequest<T> = { value?: T } & StateRequest

export default createStore
