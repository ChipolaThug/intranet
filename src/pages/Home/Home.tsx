import { FC, useEffect, memo } from "react"
import { RouteComponentProps } from "react-router-dom"
import { useDispatch, useSelector, shallowEqual } from "react-redux"
import { Helmet } from "react-helmet"

import { AppState, AppThunk } from "../../store"
import { fetchJeuxJavListIfNeed } from "../../store/jeuxJavList"
import { JeuxJavList } from "../../components"
import styles from "./styles.module.scss"

export type Props = RouteComponentProps

function useList(stateToProp: (state: AppState) => any, fetchDataIfNeed: () => AppThunk) {
    const dispatch = useDispatch()
    const { readyStatus, items } = useSelector(stateToProp, shallowEqual)

    // Fetch client-side data here
    useEffect(() => {
        dispatch(fetchDataIfNeed())
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [dispatch])

    return () => {
        if (!readyStatus || readyStatus === "invalid" || readyStatus === "request")
            return <p>Loading...</p>

        if (readyStatus === "failure") return <p>Oops, Failed to load list!</p>

        return <JeuxJavList items={items} />
    }
}

const Home: FC<Props> = (): JSX.Element => (
    <div className={styles.Home}>
        <Helmet title="Home" />
        {useList((state: AppState) => state.jeuxJavList, fetchJeuxJavListIfNeed)()}
    </div>
)

// Fetch server-side data here
export const loadData = (): AppThunk[] => [
    fetchJeuxJavListIfNeed(),
    // More pre-fetched actions...
]

export default memo(Home)
