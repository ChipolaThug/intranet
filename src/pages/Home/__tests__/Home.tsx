/**
 * @jest-environment jsdom
 */
import { render } from "@testing-library/react"
import { MemoryRouter } from "react-router-dom"

import { fetchJeuxJavListIfNeed } from "../../../store/jeuxJavList"
import mockStore from "../../../utils/mockStore"
import Home from "../Home"

describe("<Home />", () => {
    const renderHelper = (reducer = { readyStatus: "invalid" }) => {
        const { dispatch, ProviderWithStore } = mockStore({ jeuxJavList: reducer })
        const { container } = render(
            <ProviderWithStore>
                <MemoryRouter>
                    {/*
            @ts-expect-error */}
                    <Home />
                </MemoryRouter>
            </ProviderWithStore>
        )

        return { dispatch, firstChild: container.firstChild }
    }

    it("should fetch data when page loaded", () => {
        const { dispatch } = renderHelper()

        expect(dispatch).toHaveBeenCalledTimes(1)
        expect(dispatch.mock.calls[0][0].toString()).toBe(fetchJeuxJavListIfNeed().toString())
    })

    it("renders the loading status if data invalid", () => {
        expect(renderHelper().firstChild).toMatchSnapshot()
    })

    it("renders the loading status if requesting data", () => {
        const reducer = { readyStatus: "request" }

        expect(renderHelper(reducer).firstChild).toMatchSnapshot()
    })

    it("renders an error if loading failed", () => {
        const reducer = { readyStatus: "failure" }

        expect(renderHelper(reducer).firstChild).toMatchSnapshot()
    })

    it("renders the <List /> if loading was successful", () => {
        const reducer = {
            readyStatus: "success",
            items: [
                {
                    id: 5,
                    titre: "6 qui prend!",
                    auteur: "Wolfgang Kramer",
                    editeur: "(uncredited) , Design Edge , B",
                    minJoueurs: 2,
                    maxJoueurs: 10,
                    duree: 45,
                    type: "Ambiance",
                    poufpaf: "0-9-2/6-qui-prend-6-nimmt",
                    photo: "https://cf.geekdo-images.com/thumb/img/lzczxR5cw7an7tRWeHdOrRtLyes=/fit-in/200x150/pic772547.jpg",
                    bggPhoto: "",
                    bggId: 432,
                    exemplaires: 1,
                    dispoPret: 1,
                    nonRangee: 0,
                    horodatage: "0000-00-00",
                    ean: "3421272101313",
                },
            ],
        }

        expect(renderHelper(reducer).firstChild).toMatchSnapshot()
    })
})
