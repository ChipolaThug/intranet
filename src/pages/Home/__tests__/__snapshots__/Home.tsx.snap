// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> renders an error if loading failed 1`] = `
<div
  class="Home"
>
  <p>
    Oops, Failed to load list!
  </p>
</div>
`;

exports[`<Home /> renders the <List /> if loading was successful 1`] = `
<div
  class="Home"
>
  <div
    class="JeuxJavList"
  >
    <h4>
      Jeux JAV
    </h4>
    <ul>
      <li>
        6 qui prend!
         - [
        432
        ]
      </li>
    </ul>
  </div>
</div>
`;

exports[`<Home /> renders the loading status if data invalid 1`] = `
<div
  class="Home"
>
  <p>
    Loading...
  </p>
</div>
`;

exports[`<Home /> renders the loading status if requesting data 1`] = `
<div
  class="Home"
>
  <p>
    Loading...
  </p>
</div>
`;
